atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/SweepableCellsWriter.java

```diff
@@ -0,0 +1,110 @@
+
+    @Override
+    protected Map<Cell, byte[]> batchWrites(List<WriteInfo> writes) {
+        Map<Cell, byte[]> result = new HashMap<>();
+        Map<PartitionInfo, List<WriteInfo>> partitionedWrites = partitioner.filterAndPartition(writes);
+        partitionedWrites.forEach((partitionInfo, writeInfos) -> putWrites(partitionInfo, writeInfos, result));
+        return result;
+    }
+
+    private void putWrites(PartitionInfo partitionInfo, List<WriteInfo> writes, Map<Cell, byte[]> result) {
+        boolean dedicate = writes.size() > MAX_CELLS_GENERIC;
+
+        if (dedicate) {
+            addReferenceToDedicatedRows(partitionInfo, writes, result);
+        }
+
+        long index = 0;
+        for (WriteInfo write : writes) {
+            addWrite(partitionInfo, write, dedicate, index, result);
+            index++;
+        }
+    }
+
+    private void addReferenceToDedicatedRows(PartitionInfo info, List<WriteInfo> writes, Map<Cell, byte[]> result) {
+        insert(info, TableReferenceAndCell.DUMMY, false, 0, -requiredDedicatedRows(writes), result);
+    }
+
+    private void addWrite(PartitionInfo info, WriteInfo write, boolean dedicate, long index, Map<Cell, byte[]> result) {
+        insert(info, write.tableRefCell(), dedicate, index / MAX_CELLS_DEDICATED, index % MAX_CELLS_DEDICATED, result);
+    }
+
+    private void insert(PartitionInfo info, TableReferenceAndCell tableRefCell, boolean dedicate, long dedicatedRow,
+            long index, Map<Cell, byte[]> result) {
+        SweepableCellsTable.SweepableCellsRow row = createRow(info, dedicate, dedicatedRow);
+        SweepableCellsTable.SweepableCellsColumnValue colVal = createColVal(info.timestamp(), index, tableRefCell);
+        result.put(toCell(row, colVal), colVal.persistValue());
+    }
+
+    private SweepableCellsTable.SweepableCellsRow createRow(PartitionInfo info, boolean dedicate, long dedicatedRow) {
+        TargetedSweepMetadata metadata = ImmutableTargetedSweepMetadata.builder()
+                .conservative(info.isConservative().value())
+                .dedicatedRow(dedicate)
+                .shard(info.shard())
+                .dedicatedRowNumber(dedicatedRow)
+                .build();
+
+        return SweepableCellsTable.SweepableCellsRow.of(tsPartitionFine(info.timestamp()), metadata.persistToBytes());
+    }
+
+    private SweepableCellsTable.SweepableCellsColumnValue createColVal(long ts, long index,
+            TableReferenceAndCell tableRefCell) {
+        SweepableCellsTable.SweepableCellsColumn col = SweepableCellsTable.SweepableCellsColumn.of(tsMod(ts), index);
+        return SweepableCellsTable.SweepableCellsColumnValue.of(col, tableRefCell);
+    }
+
+    private long requiredDedicatedRows(List<WriteInfo> writes) {
+        return 1 + (writes.size() - 1) / MAX_CELLS_DEDICATED;
+    }
+
+    private long tsPartitionFine(long timestamp) {
+        return timestamp / TS_FINE_GRANULARITY;
+    }
+
+    public static long tsMod(long timestamp) {
```

Comment:
confirming this was intentional (seems like package visibility here would do)

Reply:
This will be refactored in a follow-up PR into a utilities class